### PR TITLE
0.7.0 release from dev to master

### PR DESCRIPTION
**Pull request by danamlewis**
0.7.0 release notes:

oref0 0.7.0 is a fairly major refactoring release that switches from the no-longer-maintained python-based `openaps` toolkit to @ecc1's Go-based libraries for pump and CGM comms.  This dramatically improves performance, particularly on the Pi Zero, and makes pump comms much more robust on all platforms.  Due to the large number of fundamental changes, extensive testing is recommended before updating existing production rigs.

**_Note for Carelink stick users:_** carelink support is being deprecated and will no longer be supported, beginning in 0.7.0 and subsequent releases. If you are a legacy carelink rig user, please install 0.6.3 from the archives. 

**_Note for TI stick users_**: 
 * There is no support for USB-connected TI sticks in 0.7.0. If you want to keep using the hardware, you'll need to jumper the header to the Pi or Edison, and re-flash the cc111x chip:
   * There is support within `oref0-setup.sh` for the TI stick when it the headers are jumper-connected to the SPI ports on the Pi (or Edison).
   * There is no support in `oref0-setup.sh` for UART-wired TI sticks in oref0-setup at the moment, but you can manually configure the software.

## Detailed notes about what's included in this release
(in approximate diff order from https://github.com/openaps/oref0/pull/1036/files):
* Use `jq` instead of `json`
* Password checking for Raspbian on initial install
* Use node 8
* Refactor into oref0-bash-common-functions.sh
* autosens-history.js (research tool)
* oref0-autotune-dayofweek.sh (for manually autotuning weekdays vs. weekends differently)
* Dev / custom branch name support in openaps-bootstrap.sh / openaps-install.sh
* Option for insulin curve tuning of insulinPeakTime and insulinEndTime on manual autotune runs
* Additional autotune field showing how often a given hour's basal was tuned vs. interpolated
* oref0-bluetoothup.sh improvements
* Refactor cron jobs into oref0-cron-every-*.sh
* Various syntax fixes and refactoring
* New oref0-g4-loop.sh using Go binaries for G4 CGM comms
* Updates to support MDT Enlite CGM (including new oref0-mdt-update.sh, oref0-monitor-cgm.sh)
* New oref0-mmtune.sh to handle mmtuning using Go pump comms
* Various oref0-online updates
* Extensive updates to oref0-pump-loop.sh for Go-based pump comms etc.
* Updates to oref0-set-device-clocks.sh to use Go-based pump comms and only update CGM time where possible
* Extensive updates to oref0-setup.sh for Go-based pump comms etc.
* Add example json files for use with oref0-simple-simulator
* Updates to autotune to improve convergence
* Support for retrospective autosens
* Support for additional treatment data entry input methods
* If CGM data is changing less than 1 mg/dL/5m for 45m, cancel any high temps and shorten any long zero temps
* Support for custom SMBInterval
* Ignore SGVs older than the last cal record to avoid dosing for calibration jumps
* Remove old `openaps` aliases (from Python code, now that we are using Go)
* Allow high_temptarget_raises_sensitivity for temptargets >= 101 and exercise_mode > 100 (vs. 111 and 105 previously)
* Support custom bolus_increment
* Update npm dependencies
* Additional unit tests
* Uploading the oref0 version and preferences file inside of devicestatus (useful for retrospective analysis)
* x12 pump users no longer need extra setup steps; this is taken care of in `oref0-setup.sh`
* xdrip-js support in `oref0`setup.sh`
* also now supports xdrip+ offline as CGM source, when rig is connected to the phone's hotspot, it can pull BGs from xdrip+ even when the phone is technically offline.

### CODE REVIEW

**Comment by tim2000s** [comment]
Had a really odd one with 0.7.0-dev today when going offline. The following happened:

```
Starting oref0-pump-loop at Mon Jun 18 16:26:02 BST 2018 with 5 second wait_for_silence:
Waiting up to 4 minutes for new BG: ..................glucose.json newer than pump_loop_completed
Listening: .No interfering pump comms detected from other rigs (this is a good thing!)
Continuing oref0-pump-loop at Mon Jun 18 16:29:13 BST 2018
Preflight OK. Profile less than 60m old; Profile valid. Pump history updated through 2018-06-18T16:17:28+01:00; meal.json ^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
Starting oref0-pump-loop at Sat Jan  1 00:01:02 GMT 2000 with 25 second wait_for_silence:
Waiting up to 4 minutes for new BG: First loop: not waiting
Listening: .No interfering pump comms detected from other rigs (this is a good thing!)
Continuing oref0-pump-loop at Sat Jan  1 00:01:31 GMT 2000
Preflight OK. Profile less than 60m old; Profile valid. Full history refresh failed. 
```
Being in mid-air there was no NTP to get a time....

**Comment by tim2000s** [comment]
That’s the odd thing. I didn’t touch the power button throughout the flight and when I landed it picked up and carried on.

**Comment by scottleibrand** [comment]
You might have a flaky power connection then.  In any event, I think the only potential software issue here is making sure it pulls the time from the pump if it can't do NTP.

**Comment by tim2000s** [comment]
Checking through the logs, I see a similar disconnect twice in the syslog, where it appears to reboot, and a third time where the clock is reset to 00:00:00 for no obvious reason. The last looks like a restart:

```
Jan  1 01:27:03 localhost systemd[642]: Failed at step NAMESPACE spawning /lib/s
ystemd/systemd-hostnamed: No such file or directory
Jan  1 01:27:03 localhost systemd[1]: systemd-hostnamed.service: main process ex
ited, code=exited, status=226/NAMESPACE
Jan  1 01:27:03 localhost systemd[1]: Failed to start Hostname Service.
Jan  1 01:27:03 localhost systemd[1]: Unit systemd-hostnamed.service entered fai
led state.
Jan  1 01:27:04 localhost dhclient: DHCPDISCOVER on wlan0 to 255.255.255.255 por
t 67 interval 6
Jan  1 01:27:04 localhost CRON[483]: (CRON) info (No MTA Jan  1 00:00:45 localho
st rsyslogd: [origin software="rsyslogd" swVersion="8.4.2" x-pid="363" x-info="h
ttp://www.rsyslog.com"] start
Jan  1 00:00:45 localhost kernel: [    0.000000] Initializing cgroup subsys cpus
et
```

**Comment by scottleibrand** [comment]
Can you help test the clock set code that is supposed to set rig time to match pump time after a power cycle while offline? It doesn’t really matter why your rig is rebooting if that works.

**Comment by tim2000s** [comment]
Yup, can do.

**Comment by tim2000s** [comment]
Haven't been able to replicate the clock issue with a basic reboot or power off. May need some further work. 

Reinstalled today with the refactored code in 0.7.0-dev on the Edison/EXB and am getting thiabconsbtly in the blogs

```
                                                                                                         
mmtune: "868.450", 0, -128 waiting for 90 second silence before continuing                               
Listening: .No interfering pump comms detected from other rigs (this is a good thing!)                   
Continuing oref0-pump-loop at Thu Jun 21 07:18:09 BST 2018                                               
                                                                                                         
Preflight fail. Done waiting for rigs with better signal.                                                
If pump and rig are close enough, this error usually self-resolves. Stand by for the next loop.          
Unsuccessful oref0-pump-loop at Thu Jun 21 07:18:43 BST 2018                                             
                                                                                                         
                                                                                                         
Starting oref0-pump-loop at Thu Jun 21 07:19:02 BST 2018 with 28 second wait_for_silence:                
Waiting up to 4 minutes for new BG: First loop: not waiting                                              
                                                                                                         
Listening: .............No interfering pump comms detected from other rigs (this is a good thing!)       
Continuing oref0-pump-loop at Thu Jun 21 07:20:58 BST 2018                                               
                                                                                                         
Preflight fail. Retry 1 of preflight                                                                     
Preflight fail. Listening: .No interfering pump comms detected from other rigs (this is a good thing!)   
Continuing oref0-pump-loop at Thu Jun 21 07:22:16 BST 2018                                               
                                                                                                         
Retry 2 of preflight                                                                                     
Preflight fail. Listening: .No interfering pump comms detected from other rigs (this is a good thing!)   
Continuing oref0-pump-loop at Thu Jun 21 07:23:21 BST 2018                                               
                                                                                                         
Retry 3 of preflight                                                                                     
Preflight fail. Couldn't preflight                                                                       
oref0-pump-loop failed. pump_loop_completed more than 15m old; waiting for 6 s silence before mmtuning   
Listening: .No interfering pump comms detected from other rigs (this is a good thing!)                   
Continuing oref0-pump-loop at Thu Jun 21 07:24:03 BST 2018                                               
                                                                                                         
Listening for 6 s silence before mmtuning: Listening: .No interfering pump comms detected from other rigs
 (this is a good thing!)                                                                                 
Continuing oref0-pump-loop at Thu Jun 21 07:24:11 BST 2018

```
Doesn't seem to have installed node 8 when running install either.

**Comment by tim2000s** [comment]
Haven’t been able to replicate the clock set issue. 

Reinstalled with the merged in refactor code on Edison and EXB and xdripaps installed fine as a data source.

**Comment by scottleibrand** [comment]
The forced upgrade to node 8 is only in the misc-stuff branch right now.

**Comment by juehv** [comment]
How do we handle this Carelink support issue?
It will not work with the current code but is still in there as option ...

**Comment by scottleibrand** [comment]
Maybe if someone chooses Carelink we tell them it's now deprecated and that they'll need to use 0.6.x if they want to use it?

**Comment by danamlewis** [comment]
(We need someone to add documentation for how to grab 0.6.x to use, so we can implement that)

**Comment by scottleibrand** [comment]
Added release notes to the top of this issue.  I think we're getting close to a release candidate for 0.7.0.  Does anyone know of anything else we need to complete before then?

File: bin/all-autosens-history.sh [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            cat entries*.json | jq -cn --stream 'fromstream(1|truncate_stream(inputs))' | while read line; do
+                date=$(echo $line | jq .dateString)
+                year=$(echo $date | cut -b 2,3,4,5)
+                month=$(echo $date | cut -b 7,8)
+                if [[ $year =~ 201. ]]; then
+                    echo "Processing participant $participant entries for $year $month"
+                    echo $line >> parts/entries-$year-$month.json
+                fi
+            done | uniq
+            for year in 2018 2017 2016; do
```
Comment (viq) [inline bin/all-autosens-history.sh]: This is missing a value now ;)
```suggestion
            for year in 2019 2018 2017 2016; do
```

**Comment by ecc1** [comment]
Here's a less fragile way to iterate over, say, the last 5 years:
cur_year=$(date +%Y)
for year in $(seq $cur_year -1 $((cur_year - 4))); do ...

**Comment by mwilliams** [comment]
Has yarn (which is seemingly blazingly fast) been looked at over NPM for dependency management?  I've been wanting to try to swap out NPM for it but life's been busy with a newborn in the house!  Mostly curious if it's been looked at or if it hasn't ever been on the radar, which is fine too!

**Comment by scottleibrand** [comment]
I've never heard of `yarn`, so it hasn't been on my radar.

**Comment by danamlewis** [comment]
- [ ] Reminder of the to-do to remove the commented section from https://github.com/openaps/oref0/pull/1244 in oref0-setup once we confirm it's not needed.

**Comment by scottleibrand** [comment]
Release notes updated (at the top, in the description of this PR).

We're getting close to being ready to finally release 0.7.0. Dana will put out a more detailed note tomorrow, but if anyone wants to get started with testing fresh installs on newly flashed rigs, and making sure there are no upstream dependency issues that we need to fix before release, that would be appreciated.

**Comment by viq** [comment]
`npm audit` is complaining for a while now about `handlebar` dependency.

**Comment by scottleibrand** [comment]
What is the complaint? Is it refusing to install, or just one of many warnings? What does it want us to do to fix?

**Comment by viq** [comment]
```
$ cd ~/src/oref0/ && git pull && git status | head -n 2 && npm install && npm audit
Already up to date.
On branch dev
Your branch is up to date with 'origin/dev'.
audited 350 packages in 1.975s
found 1 high severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details
                                                                                
                       === npm audit security report ===                        
                                                                                
┌──────────────────────────────────────────────────────────────────────────────┐
│                                Manual Review                                 │
│            Some vulnerabilities require your attention to resolve            │
│                                                                              │
│         Visit https://go.npm.me/audit-guide for additional guidance          │
└──────────────────────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ High          │ Prototype Pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ handlebars                                                   │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=4.3.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ istanbul [dev]                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ istanbul > handlebars                                        │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://npmjs.com/advisories/1164                            │
└───────────────┴──────────────────────────────────────────────────────────────┘
found 1 high severity vulnerability in 350 scanned packages
  1 vulnerability requires manual review. See the full report for details.
```

**Comment by scottleibrand** [comment]
Ok, so not a blocker. I’ll run `npm audit fix` again though.

**Comment by scottleibrand** [comment]
`npm audit fix` can't automatically fix handlebars, because istanbul, the only package that uses it inside oref0, hasn't been updated yet. We could add handlebars to our package.json, but I'd rather not force people to do a new `npm install` for that, because we only use handlebars to run tests via istanbul, not to process arbitrary input. So this vulnerability doesn't affect us, and we should probably just wait until istanbul releases a new version with the updated dependency, and `npm audit fix` again at that point.

**Comment by viq** [comment]
Roger that.

**Comment by cluckj** [comment]
`Istanbul` is deprecated and unlikely to get an update; `nyc` is the replacement but I haven't looked into what switching over would do to oref0.

**Comment by danamlewis** [comment]
Ok, we have a release candidate to begin testing in all the configurations. If you've been running dev, please update (note below that we need to test re-installs as as well as full re-flash and install from scratch methods) and observe. If you're wondering what changes are included, please see the [summarized notes](https://github.com/openaps/oref0/pull/1036#issue-195405815) at the top of the PR. After a day or two, please comment here with your setup type noting your rig type, CGM type, any particular configurations that you've tested or paid special attention to, and also your timezone. Thanks! 

- [ ] Rig types:
	- [ ] Edison & Explorer
	- [ ] Pi Zero & Explorer HAT
	- [ ] Pi 3 (?) & Explorer HAT
	- [ ] Anyone still using a TI stick that wants to test?)
	- [ ] (Any other hardware types?)
	
- [ ] Install types
	- [ ]  Re-install per dev instructions with new dev branch
	- [ ]  Reflash and install from scratch

- [ ]  All CGM types
	- [ ]  G4-Go
	- [ ]  xdrip-js methods
	- [ ]  G5
	- [ ]  G6
	- [ ]  online only via NS
	
- [ ]  Timezones
	- [ ]  Pacific
	- [ ]  Eastern
	- [ ]  Europe
	- [ ]  Australia or NZ

General notes in addition to the summary of changes above in the PR:
- After installing the version, make sure your autotune still makes sense the next few days
- Ditto for autosens
- If you're using `high_temptarget_raises_sensitivity`, make sure you test the new behavior there.

**Comment by straykatz** [comment]
@danamlewis @scottleibrand 

Oh nice!
Couple of questions re testing:
1) General: How do we generally test and report - is there a sort of form, stuff to remember? Or even some good older reports, to model on?

2) Specific to above: the boxes that are checked above - are those tests that have happened, or that are needed? (I would have thought most people run x22 and x23 ...)

Thanks, as always!

**Comment by danamlewis** [comment]
@straykatz  1. Testing is doing a new install and generally paying attention to anything in the above note about things to check (like autosens, autotune making sense etc), and generally if things are working as expected (or not). There's no form, but if you share what model pump, CGM type and method you're using, etc. when sharing if things are working as expected or not, that's how we progress with marking things off the list.  2. boxes are checked when people tell me specific setups they've tested - i.e. right now there are probably people testing with the more common models and rigs but haven't mentioned what their test setup is, so I don't know to check it off.

**Comment by straykatz** [comment]
Thank you, @danamlewis

**Comment by danamlewis** [comment]
> I like this! Could you put the public IP in the file too?

@jaylagorio won't be in time for this release, please target dev after the release to add.

**Pull request merged**